yasa/hypno.py

```diff
@@ -474,3 +477,231 @@ def hypno_find_periods(hypno, sf_hypno, threshold="5min", equal_length=False):
+            hypno[hypno == 2] = 1  # N2 becomes NREM
+            hypno[hypno == 3] = 1  # N3 becomes NREM
+            hypno[hypno == 4] = 2  # REM code changes
+    elif n_stages_out == 4:
+        # Reduce to Wake/Light/Deep/REM
+        hypno[hypno == 2] = 1  # N2 becomes Light
+        hypno[hypno == 3] = 2  # N3 becomes Deep
+        hypno[hypno == 4] = 3  # REM code changes
+    return hypno
+
+
+#############################################################################
+# SIMULATION
+#############################################################################
+
+
+def simulate_hypno(tib=90, sf=1 / 30, n_stages=5, trans_probas=None, init_probas=None, seed=None):
+    """Simulate a hypnogram based on transition probabilities.
+
+    Current implentation is a naive Markov model. The initial stage of a hypnogram
+    is generated using probabilites from ``init_probas`` and then subsequent stages
+    are generated from a Markov sequence based on ``trans_probas``.
+
+    .. important:: The Markov simulation model is not meant to accurately portray sleep
+        macroarchitecture and should only be used for testing or other unique purposes.
+
+    Parameters
+    ----------
+    tib : int
+        Time in bed, expressed in minutes.
+    sf : float
+        Sampling frequency.
+    n_stages : int
+        Staging scheme of returned hypnogram. Input should follow 5-stage scheme but can
+        be converted to lower scheme if desired.
+        .. seealso:: :py:func:`yasa.hypno_consolidate_stages`
+    trans_probas : :py:class:`pandas.DataFrame` or None
+        Transition probability matrix where each cell is a transition probability
+        between sleep stages of consecutive *epochs*.
+
+        ``trans_probas`` is a `right stochastic matrix
+        <https://en.wikipedia.org/wiki/Stochastic_matrix>`_, i.e. each row sums to 1.
+
+        If None (default), use transition probabilites from Metzner et al., 2021 [Metzner2021]_.
+        If :py:class:`pandas.DataFrame`, must have "from"-stages as indices and
+        "to"-stages as columns. Indices and columns must follow YASA integer
+        hypnogram convention (W = 0, N = 1, ...). Unscored/Artefact stages are not allowed.
+
+        .. note:: Transition probability matrices should indicate the transition
+            probability between *epochs* (i.e., probability of the next epoch) and
+            not simply stage (i.e., probability of non-similar stage).
+
+        .. seealso:: Return value from :py:func:`yasa.transition_matrix`
+    init_probas : :py:class:`pandas.Series` or None
+        Probabilites of each stage to initialize random walk.
+        If None (default), initialize with "From"-Wake row of ``trans_probas``.
+        If :py:class:`pandas.Series`, indices must be stages following YASA integer
+        hypnogram convention (see ``trans_probas``).
+    seed : int or None
+        Random seed for generating Markov sequence.
```

Comment:
Can you please add that a) this can be any integer number and b) this is required to get reproducible results? I've seen that many people don't really understand the concept of seed

Reply:
✅ In updating this I noticed that I'm using a legacy approach to using random seeds in numpy. I think the updated approach is better (and doesn't even need to load in `scipy.stats`). Lmk if you see a problem with it.